postgres/datadog_checks/postgres/statements.py

```diff
@@ -95,32 +91,47 @@ def _get_pg_stat_statements_columns(self, db):
         version is not a reliable way to determine the available columns on `pg_stat_statements`. The database can
         be upgraded without upgrading extensions, even when the extension is included by default.
         """
+        if len(self._stat_column_cache) == 0:
+            stat_column_query = PG_STAT_STATEMENTS_COLUMN_QUERY.format(
+                pg_stat_statements_view=self._config.pg_stat_statements_view
+            )
+            stat_column_cursor = db.cursor()
+            self._execute_query(stat_column_cursor, stat_column_query, params=(self._config.dbname,))
+            if not stat_column_cursor.description:
+                self._log.exception("Failed to query pg stat statement columns")
+            self._stat_column_cache = set([desc[0] for desc in stat_column_cursor.description])
+
         # Querying over '*' with limit 0 allows fetching only the column names from the cursor without data
-        query = STATEMENTS_QUERY.format(
+        statements_query = STATEMENTS_QUERY.format(
```

Comment:
Is this second query necessary given that we already made the `PG_STAT_STATEMENTS_COLUMN_QUERY` above?

Reply:
@djova I believe we do, the second query where we `JOIN pg_roles` and `JOIN pg_database` in `STATEMENTS_QUERY` allows us to retrieve the [`datname` and `rolname`](https://github.com/DataDog/integrations-core/blob/master/postgres/datadog_checks/postgres/statements.py#L163) which we need to compute the derivative rows. The only intention of querying just the stat statement columns and caching it was to use a diff to  get the columns we want (i.e we don't want columns of type string). I still like this idea of caching just the columns we want and using it as a diff since the alternative solution is to explicitly define the columns we don't want in the block list which we'd have to maintain where as the we don't have to with the cache & diff approach.

Reply:
Oh right it's used in the following query. One option is to hardcode those couple of columns (`datname` and `rolname`) (since they're required anyway) and just query for them regardless. If for some reason they're missing then the query will fail and the end-result will be the same anyway, where we have no metrics. 

That way `_get_pg_stat_statements_columns` can be only for the `pg_stat_statements` columns which is the part that needs to be dynamic.

Reply:
On second thought, maybe the allow list approach [discussed here](https://github.com/DataDog/integrations-core/pull/9365#discussion_r634570939) will be simpler after all. In that case we could just keep the same logic as before here and simply add caching to avoid redoing the query every time.